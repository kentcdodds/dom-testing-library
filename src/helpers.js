const globalObj = typeof window === 'undefined' ? global : window
// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

// Currently this fn only supports jest timers, but it could support other test runners in the future.
function runWithRealTimers(callback) {
  const fakeTimersType = getJestFakeTimersType()
  if (fakeTimersType) {
    jest.useRealTimers()
  }

  const callbackReturnValue = callback()

  if (fakeTimersType) {
    jest.useFakeTimers(fakeTimersType)
  }

  return callbackReturnValue
}

function getJestFakeTimersType() {
  // istanbul ignore if
  if (
    typeof jest === 'undefined' ||
    typeof globalObj.setTimeout === 'undefined'
  ) {
    return null
  }

  if (
    typeof globalObj.setTimeout._isMockFunction !== 'undefined' &&
    globalObj.setTimeout._isMockFunction
  ) {
    return 'legacy'
  }

  if (
    typeof globalObj.setTimeout.clock !== 'undefined' &&
    typeof jest.getRealSystemTime !== 'undefined'
  ) {
    try {
      // jest.getRealSystemTime is only supported for Jest's `modern` fake timers and otherwise throws
      jest.getRealSystemTime()
      return 'modern'
    } catch {
      // not using Jest's modern fake timers
    }
  }
  return null
}

const jestFakeTimersAreEnabled = () => Boolean(getJestFakeTimersType())

// we only run our tests in node, and setImmediate is supported in node.
// istanbul ignore next
function setImmediatePolyfill(fn) {
  return globalObj.setTimeout(fn, 0)
}

function getTimeFunctions() {
  // istanbul ignore next
  return {
    clearTimeoutFn: globalObj.clearTimeout,
    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
    setTimeoutFn: globalObj.setTimeout,
  }
}

const {clearTimeoutFn, setImmediateFn, setTimeoutFn} = runWithRealTimers(
  getTimeFunctions,
)

function getDocument() {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(node) {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument && node.ownerDocument.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (node.then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `Unable to find the "window" object for the given node. Please file an issue with the code that's causing you to see this error: https://github.com/testing-library/dom-testing-library/issues/new`,
    )
  }
}

/**
 * Check if an element is of a given type.
 *
 * @param Element The element to test
 * @param string Constructor name. E.g. 'HTMLSelectElement'
 */
function isInstanceOfElement(element, elementType) {
  try {
    const window = getWindowFromNode(element)
    // Window usually has the element constructors as properties but is not required to do so per specs
    if (typeof window[elementType] === 'function') {
      return element instanceof window[elementType]
    }
  } catch (e) {
    // The document might not be associated with a window
  }

  // Fall back to the constructor name as workaround for test environments that
  // a) not associate the document with a window
  // b) not provide the constructor as property of window
  if (/^HTML(\w+)Element$/.test(element.constructor.name)) {
    return element.constructor.name === elementType
  }

  // The user passed some node that is not created in a browser-like environment
  throw new Error(
    `Unable to verify if element is instance of ${elementType}. Please file an issue describing your test environment: https://github.com/testing-library/dom-testing-library/issues/new`,
  )
}

function checkContainerType(container) {
  if (
    !container ||
    !(typeof container.querySelector === 'function') ||
    !(typeof container.querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  clearTimeoutFn as clearTimeout,
  setImmediateFn as setImmediate,
  setTimeoutFn as setTimeout,
  runWithRealTimers,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
  isInstanceOfElement,
}
