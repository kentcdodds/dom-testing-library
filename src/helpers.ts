import {Screen} from '../types'

const globalObj = typeof window === 'undefined' ? global : window
// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

// Currently this fn only supports jest timers, but it could support other test runners in the future.
function runWithRealTimers<T>(callback: () => T): T {
  return hasJestTimers()
    ? runWithJestRealTimers(callback).callbackReturnValue
    : // istanbul ignore next
      callback()
}

function hasJestTimers(): boolean {
  return (
    typeof jest !== 'undefined' &&
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    jest !== null &&
    typeof jest.useRealTimers === 'function'
  )
}

export interface TimerApi {
  clearInterval: typeof clearInterval
  clearTimeout: typeof clearTimeout
  setInterval: typeof setInterval
  setTimeout: typeof setTimeout & {clock?: unknown}
  setImmediate?: typeof setImmediate
  clearImmediate?: typeof clearImmediate
}

function runWithJestRealTimers<T>(callback: () => T) {
  const timerAPI: TimerApi = {
    clearInterval,
    clearTimeout,
    setInterval,
    setTimeout,
  }

  // For more on why we have the check here,
  // checkout https://github.com/testing-library/dom-testing-library/issues/914
  if (typeof setImmediate === 'function') {
    timerAPI.setImmediate = setImmediate
  }
  if (typeof clearImmediate === 'function') {
    timerAPI.clearImmediate = clearImmediate
  }

  jest.useRealTimers()

  const callbackReturnValue = callback()

  const usedFakeTimers = Object.entries(timerAPI).some(
    ([name, func]) => func !== globalObj[name as keyof TimerApi],
  )

  if (usedFakeTimers) {
    // eslint-disable-next-line  @typescript-eslint/no-unnecessary-condition
    jest.useFakeTimers(timerAPI.setTimeout?.clock ? 'modern' : 'legacy')
  }

  return {
    callbackReturnValue,
    usedFakeTimers,
  }
}

function jestFakeTimersAreEnabled(): boolean {
  return hasJestTimers()
    ? runWithJestRealTimers(() => {}).usedFakeTimers
    : // istanbul ignore next
      false
}

// we only run our tests in node, and setImmediate is supported in node.
// istanbul ignore next
function setImmediatePolyfill(fn: TimerHandler): number {
  return globalObj.setTimeout(fn, 0)
}

function getTimeFunctions() {
  // istanbul ignore next
  return {
    clearTimeoutFn: globalObj.clearTimeout,
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
    setTimeoutFn: globalObj.setTimeout,
  }
}

const {clearTimeoutFn, setImmediateFn, setTimeoutFn} = runWithRealTimers(
  getTimeFunctions,
)

function getDocument(): Document {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(
  node: EventTarget & {
    defaultView?: (Window & typeof globalThis) | null
    ownerDocument?: Document | null
    window?: Window & typeof globalThis
  },
): Window & typeof globalThis {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument?.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (((node as unknown) as Promise<unknown>).then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else if (Array.isArray(node)) {
    throw new Error(
      `It looks like you passed an Array instead of a DOM node. Did you do something like \`fireEvent.click(screen.getAllBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`?`,
    )
  } else if (
    typeof ((node as unknown) as Screen).debug === 'function' &&
    typeof ((node as unknown) as Screen).logTestingPlaygroundURL === 'function'
  ) {
    throw new Error(
      `It looks like you passed a \`screen\` object. Did you do something like \`fireEvent.click(screen, ...\` when you meant to use a query, e.g. \`fireEvent.click(screen.getBy..., \`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `Unable to find the "window" object for the given node. Please file an issue with the code that's causing you to see this error: https://github.com/testing-library/dom-testing-library/issues/new`,
    )
  }
}

export type Container = Document | DocumentFragment | Element

function checkContainerType(
  container: Node | null,
): asserts container is Container {
  if (
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    !container ||
    !(typeof (container as Container).querySelector === 'function') ||
    !(typeof (container as Container).querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object: unknown) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  clearTimeoutFn as clearTimeout,
  setImmediateFn as setImmediate,
  setTimeoutFn as setTimeout,
  runWithRealTimers,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
}
