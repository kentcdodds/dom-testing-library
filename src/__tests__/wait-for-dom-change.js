import {waitForDomChange} from '../'
// adds special assertions like toBeTruthy
import 'jest-dom/extend-expect'
import {render} from './helpers/test-utils'

async function skipSomeTime(delayMs) {
  await new Promise(resolve => setTimeout(resolve, delayMs))
}

async function skipSomeTimeForMutationObserver(delayMs = 50) {
  // Using `setTimeout` >30ms instead of `wait` here because `mutationobserver-shim` uses `setTimeout` ~30ms.
  await skipSomeTime(delayMs, 50)
}

test('it waits for the next DOM mutation', async () => {
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForDomChange().then(successHandler, errorHandler)

  // Promise callbacks are always asynchronous.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  await skipSomeTimeForMutationObserver()

  // No more expected calls without DOM mutations.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  document.body.appendChild(document.createElement('div'))
  expect(document.body).toMatchSnapshot()

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledWith(true)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(document.body).toMatchSnapshot()

  return promise
})

test('it waits characterData mutation', async () => {
  const {container} = render(`<div>initial text</div>`)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForDomChange({container}).then(
    successHandler,
    errorHandler,
  )

  // Promise callbacks are always asynchronous.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(container).toMatchSnapshot()

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.querySelector('div').innerHTML = 'new text'
  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(container).toMatchSnapshot()

  return promise
})

test('it waits for the attributes mutation', async () => {
  const {container} = render(``)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForDomChange({
    container,
  }).then(successHandler, errorHandler)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'PASSED')
  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledWith(true)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  return promise
})

test('it throws if timeout is exceeded', async () => {
  const {container} = render(``)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForDomChange({
    container,
    timeout: 70,
    mutationObserverOptions: {attributes: true},
  }).then(successHandler, errorHandler)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // no DOM update made once
  await skipSomeTimeForMutationObserver(50)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // no DOM update made twice
  await skipSomeTimeForMutationObserver(50)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler.mock.calls[0]).toMatchSnapshot()
  expect(container).toMatchSnapshot()
  return promise
})

test('does not get into infinite setTimeout loop after MutationObserver notification', async () => {
  const {container} = render(`<div data-testid="initial-element"></div>`)

  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')
  jest.useFakeTimers()

  const promise = waitForDomChange({
    container,
    timeout: 70,
    mutationObserverOptions: {attributes: true},
  }).then(successHandler, errorHandler)

  // Expect 2 timeouts to be scheduled:
  // - waitForDomChange timeout
  // - MutationObserver timeout
  expect(setTimeout).toHaveBeenCalledTimes(2)

  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // Make a mutation
  container.setAttribute('data-test-attribute', 'something changed')

  // Advance timer to expire MutationObserver timeout
  jest.advanceTimersByTime(50)
  jest.runAllImmediates()
  await promise
  expect(setTimeout).toHaveBeenCalledTimes(3)

  // Expect successHandler to be called
  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // Expect no more setTimeout calls
  jest.advanceTimersByTime(100)
  expect(setTimeout).toHaveBeenCalledTimes(3)
  jest.useRealTimers()
})
