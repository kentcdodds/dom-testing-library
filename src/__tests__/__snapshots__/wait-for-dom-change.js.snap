// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it throws if timeout is exceeded 1`] = `
Array [
  [Error: Timed out in waitForDomChange.],
]
`;

exports[`it throws if timeout is exceeded 2`] = `<div />`;

exports[`it waits characterData mutation 1`] = `
<div>
  <div>
    initial text
  </div>
</div>
`;

exports[`it waits characterData mutation 2`] = `
<div>
  <div>
    new text
  </div>
</div>
`;

exports[`it waits for the next DOM mutation 1`] = `
<body>
  <div />
</body>
`;

exports[`it waits for the next DOM mutation 2`] = `
<body>
  <div />
</body>
`;
